structured_data/2021_10_01_imbalanced_datasets/README.md

```diff
@@ -0,0 +1,130 @@
+| Prototype generation (generate samples based on majority class data) | *ClusterCentroids* | Selecting the centroids of K-means                                                                             |
+
+Illustration of Tomek's links:
+
+![Illustration of Tomek's links](./figures/tomeks_links.png)
+
+Cluster centroids example with [imbalanced-learn](https://pypi.org/project/imbalanced-learn/) (previously `imblearn`):
+
+```python
+from collections import Counter
+
+from sklearn.datasets import make_classification
+from imblearn.under_sampling import ClusterCentroids
+
+X, y = make_classification(n_samples=5000, n_features=2, n_redundant=0,
+                          n_classes=3, n_clusters_per_class=1,
+                          weights=[0.01, 0.05, 0.94], random_state=0)
+print(sorted(Counter(y).items())) # [(0, 64), (1, 262), (2, 4674)]
+X_resampled, y_resampled = ClusterCentroids().fit_resample(X, y)
+print(sorted(Counter(y_resampled).items())) # [(0, 64), (1, 64), (2, 64)]
+```
+
+The following image shows the dataset before and after resampling:
+
+![Cluster centroids imblearn](./figures/clustercentroids.png)
+
+## 3. Over-sample the minority class
+
+This can be done in 3 different ways:
+
+* **Random sampling**: pretty straightforward.
+
+* **SMOTE**: Generate synthetic data by interpolating between original data points. (There are many variants available; check the [documentation](https://imbalanced-learn.org/stable/over_sampling.html#smote-variants)! )
+
+```python
+from collections import Counter
+
+from sklearn.datasets import make_classification
+from imblearn.over_sampling import SMOTE
+
+X, y = make_classification(n_samples=5000, n_features=2, n_redundant=0,
+                          n_clusters_per_class=1, weights=[0.01, 0.99],
+                          random_state=0)
+print(sorted(Counter(y).items())) # [(0, 77), (1, 4923)]
+X_resampled, y_resampled = SMOTE().fit_resample(X, y)
+print(sorted(Counter(y_resampled).items())) # [(0, 4923), (1, 4923)]
+```
+
+* **ADASYN**: Same as SMOTE but puts more focus on harder to learn-neighborhoods by selecting samples that are wrongly classified using KNN.
+
+![Over-sampling with SMOTE and ADASYN](./figures/oversampling.png)
+
+## 4. Be careful for data leakage
+
+**You should separate training and testing data before re-sampling.**
+
+* Obvious reason: Wielding generative sampling approaches, you might end up with training data that is generated from test data.
+
+* Not-so-obvious reason: Training data is sampled because we want to help the model better assess all labels. If we resample the test data too, we might digress from reality and get to a wrong/biased performance assessment.
+
+Pro-tip: Use [sklearn Pipelines](https://imbalanced-learn.org/stable/references/generated/imblearn.pipeline.Pipeline.html) to conveniently keep an overview of the steps and minimize the risk of mistakes and data leakage when sampling.
+
+## 5. Use built-in model functionality
```

Comment:
Might be interesting to link the documentation where we got this information from